src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/aggregators/PersistableSlidingWindow.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ *  permissions and limitations under the License.
+ */
+
+package com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.aggregators;
+
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.persistence.FileRotate;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import java.io.BufferedWriter;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.Writer;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.concurrent.TimeUnit;
+import org.apache.commons.io.FileUtils;
+import org.apache.commons.io.LineIterator;
+import org.apache.commons.lang3.RandomStringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+/**
+ * PersistableSlidingWindow is a SlidingWindow which can have its data written to and read from disk
+ *
+ * <p>The sliding window retains data for slidingWindowSizeInSeconds and aggregates values into time
+ * buckets of width bucketSizeInSeconds. For example if slidingWindowSizeInSeconds is 86400 (1 day)
+ * and bucketSizeInSeconds is 3600 (1 hr) then there will be at most 25 data points in the sliding window
+ * at any given time. It's 25 and not 24 because the oldest datapoints are removed lazily based on
+ * arrival time. This shouldn't impact most consumers of this data structure.
+ *
+ * <p>The data contained in this sliding window is effectively a time series. Consumers may extend this
+ * class and define more complex aggregations or analytics functions.
+ */
+public class PersistableSlidingWindow extends SlidingWindow<SlidingWindowData> {
```

Comment:
This class is doing two things, it is providing a persistable sliding window, and it is aggregating (deduping) data points within a bucketization window.
How about we split these responsibilities and have this class focus only on persisting?

Re: the bucketization piece, I have a few more comments - I suppose the motivation for this is to act like an `alarm breach monitor` which gets set when some thresholds are hit, and unset on a different criteria. This can be used in deciders then, to figure out when an action is relevant (alarm unhealthy) and irrelevant (alarm healthy).

As such, it may be read more often than an issue gets added (`next` being called) and the sliding window should get updated on each read. 

Again, we can move this to a separate PR/class and focus this one primarily on persistence as I see most of the code here pertains to persisting data.

Reply:
I removed the aggregation logic from this PR.